apis/server/image_bridge.go

```diff
@@ -207,3 +207,28 @@ func (s *Server) getImageHistory(ctx context.Context, rw http.ResponseWriter, re
 
 	return EncodeResponse(rw, http.StatusOK, history)
 }
+
+// pushImage will push an image to a specified registry.
+func (s *Server) pushImage(ctx context.Context, rw http.ResponseWriter, req *http.Request) error {
+	name := mux.Vars(req)["name"]
+	tag := req.FormValue("tag")
+
+	// get registry auth from Request header
+	authStr := req.Header.Get("X-Registry-Auth")
+	authConfig := types.AuthConfig{}
+	if authStr != "" {
+		data := base64.NewDecoder(base64.URLEncoding, strings.NewReader(authStr))
+		if err := json.NewDecoder(data).Decode(&authConfig); err != nil {
+			return err
+		}
+	}
+
+	// Error information has be sent to client, so no need call resp.Write
+	// no need return error here
+	if err := s.ImageMgr.PushImage(ctx, name, tag, &authConfig, newWriteFlusher(rw)); err != nil {
+		logrus.Errorf("failed to push image %s with tag %s: %v", name, tag, err)
+		return nil
```

Comment:
`return err` or `return nil`?

Reply:
I think like `pull` and `exec`, return nil here

Reply:
could we check image before push? `PushImage` will return 404 as error. And the user doesn't know error because we don't return it.

Reply:
@fuweid , have do the check, if containerd can get image, we will push

Reply:
I re-check the code here, we use jsonstream to pass infos to client, but stream only start after push image, that means no other error can not be send to client, so I think should be `return error` here